secure_fw/partitions/internal_trusted_storage/its_crypto_interface.c

```diff
@@ -0,0 +1,186 @@
+ *                                    buffers are NULL
+ *
+ */
+static psa_status_t tfm_its_fill_enc_add(uint8_t *add,
+                                         size_t add_size,
+                                         uint8_t *fid,
+                                         size_t fid_size,
+                                         uint32_t flags,
+                                         size_t data_size)
+
+{
+    /* Only the user flags are populated in the function which
+     * gets the file info from ITS (see its_flash_fs_file_get_info).
+     * We use the same flags for comformity.
+     */
+    uint32_t user_flags = flags & ITS_FLASH_FS_USER_FLAGS_MASK;
+
+    /* The additional data consist of the file id, the flags and the
+     * data size of the file.
+     */
+    size_t add_expected_size = ITS_FILE_ID_SIZE +
+                               sizeof(user_flags) +
+                               sizeof(data_size);
+
+    if (add_size != add_expected_size || add == NULL || fid == NULL) {
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+
+    memcpy(add, fid, fid_size);
+    memcpy(add + fid_size, &user_flags, sizeof(user_flags));
+    memcpy(add + fid_size + sizeof(user_flags),
+               &data_size,
+               sizeof(data_size));
+
+    return PSA_SUCCESS;
+}
+
+static psa_status_t tfm_hal_to_psa_error(enum tfm_hal_status_t tfm_hal_err){
+
+    switch (tfm_hal_err)
+    {
+    case TFM_HAL_SUCCESS:
+        return PSA_SUCCESS;
+        break;
+    case TFM_HAL_ERROR_INVALID_INPUT:
+        return PSA_ERROR_INVALID_ARGUMENT;
+        break;
+    default:
+        return PSA_ERROR_GENERIC_ERROR;
+        break;
+    }
+
+}
+
+psa_status_t tfm_its_crypt_file(struct its_flash_fs_file_info_t *finfo,
+                                uint8_t *fid,
+                                size_t fid_size,
+                                uint8_t *input,
+                                size_t input_size,
+                                uint8_t *output,
+                                size_t output_size,
+                                bool is_encrypt)
+{
+
+    struct tfm_hal_its_auth_crypt_ctx aead_ctx = {0};
+    enum tfm_hal_status_t err;
+    size_t file_size;
+
+    if (finfo == NULL) {
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+
+    /* The file size is not known yet when encrypting */
+    if (is_encrypt){
+        file_size = input_size;
+    } else {
+        file_size = finfo->size_current;
+    }
+
+    err =  tfm_its_fill_enc_add(finfo->add,
+                                sizeof(finfo->add),
+                                fid,
+                                fid_size,
+                                finfo->flags,
+                                file_size);
+    if (err != TFM_HAL_SUCCESS) {
+        return tfm_hal_to_psa_error(err);
+    }
+
+    if (is_encrypt) {
+        err = tfm_hal_its_aead_generate_nonce(finfo->nonce,
+                                              sizeof(finfo->nonce));
```

Comment:
Again @Vge0rge eventually can give an explanation why this approach was chosen. 
Without looking into it I wonder what happens if all ITS files get deleted. How would we then prevent starting from 0 again?

Could also be that I don't understand the suggestions correctly.

Reply:
That is a good point.

I had another idea.

What about using the file ID as the nonce?

I am unable to think of any problems with this approach.

Reply:
Hmm, actually, I wonder if this might break the unique requirement somehow ...

Reply:
Yeah, I think it has the same problem as you identified for a montonic counter. When you wipe the filesystem you can encrypt a different plaintext with the same nonce.

Reply:
According to RFC 5116 g_enc_nonce_seed must be fixed across resets and unique per device. Presumably to avoid the birthday collision problem.

"The Fixed field MUST remain constant for all nonces that are generated for a given encryption device."

https://www.rfc-editor.org/rfc/rfc5116#section-3.1

Our current approach appears to be violating a MUST statement of the RFC and therefore be unsecure unless I am missing something.

Reply:
I double-checked this, and it's not the same scheme, our counter isn't monotonic.

Do we have a trustworthy source, like an RFC, that says our approach to the the nonce is secure?

Reply:
After offline discussion added a disclaimer at the  beginning of `tfm_hal_its_aead_generate_nonce`

Anything else will be handled in the updated documentation afterwards